File: lib/external-gtfs/frequency-source.js
'use strict';
/**
 * Source « fréquence documentée » : produit un statut ESTIMATED (intervalle de
 * passage publié par une source identifiée), JAMAIS une heure de passage.
 * Elle n'intervient qu'après toutes les sources statiques actuelles
 * (stop_times) et uniquement pour la ligne et la période documentées.
 *
 * Aucune fréquence n'est embarquée par défaut : les entrées doivent provenir
 * d'une publication identifiée (source, url, published_at, validité).
 */
const { parseTime, normalizeServiceDate, toIsoDate, formatTime, weekdayOf } = require('./gtfs-time');
const { FeedProvenance } = require('./provenance');
const { STATUS, REASONS } = require('./gtfs-schedule-service');

class FrequencySource {
  /**
   * @param {Array<object>} entries [{ network, lineNumber, routeIds?, headwayMinutes, days?: ['monday',…], from?: 'HH:MM', to?: 'HH:MM', label? }]
   * @param {FeedProvenance|object} provenance provenance de la publication (source, url, published_at, valid_from, valid_to)
   */
  constructor(entries, provenance) {
    this.provenance = provenance instanceof FeedProvenance ? provenance : new FeedProvenance(provenance);
    this.entries = (entries || []).map((e, k) => {
      if (!e || !e.network || !e.lineNumber) throw new Error(`FrequencySource : entrée ${k} sans network / lineNumber`);
      const headway = Number(e.headwayMinutes);
      if (!Number.isFinite(headway) || headway <= 0) throw new Error(`FrequencySource : headwayMinutes invalide (${e.lineNumber})`);
      return {
        network: String(e.network).toUpperCase(),
        lineNumber: String(e.lineNumber).trim().toUpperCase(),
        routeIds: Array.isArray(e.routeIds) ? e.routeIds.slice() : [],
        headwayMinutes: headway,
        days: Array.isArray(e.days) && e.days.length ? e.days.map((d) => String(d).toLowerCase()) : null,
        fromSeconds: e.from ? parseTime(e.from) : null,
        toSeconds: e.to ? parseTime(e.to) : null,
        label: e.label || null,
      };
    });
    this.network = this.entries.length ? this.entries[0].network : (this.provenance.network || null);
    this.kind = 'frequency';
  }

  /** Pseudo-lignes : numéro public tel que publié (jamais déduit). */
  routes() {
    return this.entries.map((e) => ({
      routeId: `FREQ:${e.network}:${e.lineNumber}`,
      routeShortName: e.lineNumber,
      routeLongName: e.label,
      routeType: null,
      agencyId: null,
      network: e.network,
      raw: {},
    }));
  }

  stops() { return []; }
  knowsRoute(routeId) { return this.entries.some((e) => e.routeIds.includes(routeId) || `FREQ:${e.network}:${e.lineNumber}` === routeId); }

  /** Entrée applicable à une ligne, une date et une heure. */
  entryFor(network, lineNumber, date, time) {
    const net = network ? String(network).toUpperCase() : null;
    const line = String(lineNumber || '').trim().toUpperCase();
    const day = normalizeServiceDate(date);
    const weekday = weekdayOf(day);
    const seconds = parseTime(time);
    for (const e of this.entries) {
      if (net && e.network !== net) continue;
      if (e.lineNumber !== line) continue;
      if (e.days && !e.days.includes(weekday)) continue;
      if (seconds !== null && e.fromSeconds !== null && seconds < e.fromSeconds) continue;
      if (seconds !== null && e.toSeconds !== null && seconds > e.toSeconds) continue;
      return e;
    }
    return null;
  }

  /**
   * Interface commune : statut ESTIMATED pour la ligne (l'arrêt n'est pas
   * vérifiable dans une simple fréquence, ce qui est dit explicitement).
   */
  getDeparturesAtStop(routeId, stopId, date, time, options = {}) {
    const serviceDate = normalizeServiceDate(date);
    const asOf = options.asOf ? normalizeServiceDate(options.asOf) : serviceDate;
    const nowSeconds = parseTime(time);
    const entry = this.entries.find((e) => e.routeIds.includes(routeId) || `FREQ:${e.network}:${e.lineNumber}` === routeId) || null;
    const applicable = entry ? this.entryFor(entry.network, entry.lineNumber, serviceDate, time) : null;
    const base = {
      routeId,
      stopId,
      date: toIsoDate(serviceDate),
      currentTime: nowSeconds === null ? null : formatTime(nowSeconds),
      source: this.provenance.source,
      sourceType: this.provenance.sourceTypeLabel,
      authority: this.provenance.authority,
      feedVersion: this.provenance.feedVersion,
      validity: this.provenance.validityOn(asOf, serviceDate),
      provenanceLevel: this.provenance.isCurrentOn(asOf) ? 'ESTIMATED' : 'HISTORICAL_REFERENCE',
      isCurrent: false,
      departures: [],
    };
    if (nowSeconds === null) return { ...base, status: STATUS.UNKNOWN, reason: REASONS.INVALID_TIME };
    if (!entry) return { ...base, status: STATUS.UNKNOWN, reason: REASONS.ROUTE_UNKNOWN };
    if (!applicable) return { ...base, status: STATUS.UNKNOWN, reason: REASONS.NO_SERVICE_ON_DATE };
    return {
      ...base,
      status: STATUS.ESTIMATED,
      reason: null,
      estimate: {
        network: applicable.network,
        lineNumber: applicable.lineNumber,
        headwayMinutes: applicable.headwayMinutes,
        window: {
          from: applicable.fromSeconds === null ? null : formatTime(applicable.fromSeconds),
          to: applicable.toSeconds === null ? null : formatTime(applicable.toSeconds),
          days: applicable.days,
        },
        stopVerified: false,
        source: this.provenance.source,
        url: this.provenance.url,
        publishedAt: this.provenance.publishedAt,
      },
    };
  }
}

module.exports = { FrequencySource };

File: flutter-src/lib/services/external_gtfs/frequency_source.dart
// Source « fréquence documentée » — miroir Dart de lib/external-gtfs/frequency-source.js.
// Produit un statut ESTIMATED (intervalle publié par une source identifiée),
// JAMAIS une heure de passage. Aucune fréquence n'est embarquée par défaut.

import 'feed_provenance.dart';
import 'gtfs_feed.dart';
import 'gtfs_schedule_service.dart';
import 'gtfs_time.dart';

class FrequencyEntry {
  FrequencyEntry({
    required String network,
    required String lineNumber,
    required this.headwayMinutes,
    List<String>? routeIds,
    List<String>? days,
    String? from,
    String? to,
    this.label,
  })  : network = network.toUpperCase(),
        lineNumber = lineNumber.trim().toUpperCase(),
        routeIds = routeIds ?? <String>[],
        days = days == null || days.isEmpty
            ? null
            : days.map((String d) => d.toLowerCase()).toList(),
        fromSeconds = from == null ? null : parseGtfsTime(from),
        toSeconds = to == null ? null : parseGtfsTime(to) {
    if (headwayMinutes <= 0) {
      throw ArgumentError('FrequencyEntry : headwayMinutes invalide ($lineNumber)');
    }
  }

  final String network;
  final String lineNumber;
  final int headwayMinutes;
  final List<String> routeIds;
  final List<String>? days;
  final int? fromSeconds;
  final int? toSeconds;
  final String? label;

  String get pseudoRouteId => 'FREQ:$network:$lineNumber';
}

class FrequencySource implements ScheduleSource {
  FrequencySource(this.entries, this.provenance);

  final List<FrequencyEntry> entries;

  @override
  final FeedProvenance provenance;

  @override
  String? get network => entries.isEmpty ? provenance.network : entries.first.network;

  @override
  String get kind => 'frequency';

  @override
  List<GtfsRoute> routes() => entries
      .map((FrequencyEntry e) => GtfsRoute(
            routeId: e.pseudoRouteId,
            routeShortName: e.lineNumber,
            routeLongName: e.label ?? '',
            routeType: '',
            agencyId: null,
          ))
      .toList();

  @override
  List<GtfsStop> stops() => <GtfsStop>[];

  @override
  List<RouteDirectionStops> getStopsForRoute(String routeId) => <RouteDirectionStops>[];

  @override
  bool knowsRoute(String routeId) => _entryByRoute(routeId) != null;

  FrequencyEntry? _entryByRoute(String routeId) {
    for (final FrequencyEntry e in entries) {
      if (e.routeIds.contains(routeId) || e.pseudoRouteId == routeId) return e;
    }
    return null;
  }

  /// Entrée applicable à une ligne, une date et une heure.
  FrequencyEntry? entryFor(String? network, String lineNumber, String date, String time) {
    final String? net = network?.toUpperCase();
    final String line = lineNumber.trim().toUpperCase();
    final String day = normalizeServiceDate(date);
    final String weekday = weekdayOf(day);
    final int? seconds = parseGtfsTime(time);
    for (final FrequencyEntry e in entries) {
      if (net != null && e.network != net) continue;
      if (e.lineNumber != line) continue;
      if (e.days != null && !e.days!.contains(weekday)) continue;
      if (seconds != null && e.fromSeconds != null && seconds < e.fromSeconds!) continue;
      if (seconds != null && e.toSeconds != null && seconds > e.toSeconds!) continue;
      return e;
    }
    return null;
  }

  @override
  DepartureResult getDeparturesAtStop(
    String routeId,
    String? stopId,
    String date,
    String time, {
    int? limit = 5,
    String? asOf,
  }) {
    final String serviceDate = normalizeServiceDate(date);
    final String ref = asOf == null ? serviceDate : normalizeServiceDate(asOf);
    final int? nowSeconds = parseGtfsTime(time);
    final FrequencyEntry? entry = _entryByRoute(routeId);
    final FrequencyEntry? applicable =
        entry == null ? null : entryFor(entry.network, entry.lineNumber, serviceDate, time);
    final String level = provenance.isCurrentOn(ref)
        ? ProvenanceLevels.estimated
        : ProvenanceLevels.historicalReference;
    String status = ScheduleStatus.unknown;
    String? reason;
    if (nowSeconds == null) {
      reason = ScheduleReasons.invalidTime;
    } else if (entry == null) {
      reason = ScheduleReasons.routeUnknown;
    } else if (applicable == null) {
      reason = ScheduleReasons.noServiceOnDate;
    } else {
      status = ScheduleStatus.estimated;
    }
    return DepartureResult(
      routeId: routeId,
      stopId: stopId,
      date: toIsoDate(serviceDate),
      currentTime: nowSeconds == null ? null : formatGtfsTime(nowSeconds),
      status: status,
      reason: reason,
      source: provenance.source,
      sourceType: provenance.sourceTypeLabel,
      feedVersion: provenance.feedVersion,
      validity: provenance.validityOn(ref, serviceDate),
      provenanceLevel: level,
      departures: const <Departure>[],
      network: network,
      estimate: status == ScheduleStatus.estimated && applicable != null
          ? FrequencyEstimate(
              network: applicable.network,
              lineNumber: applicable.lineNumber,
              headwayMinutes: applicable.headwayMinutes,
              source: provenance.source,
              windowFrom: applicable.fromSeconds == null
                  ? null
                  : formatGtfsTime(applicable.fromSeconds!),
              windowTo: applicable.toSeconds == null
                  ? null
                  : formatGtfsTime(applicable.toSeconds!),
              url: provenance.url,
              publishedAt: provenance.publishedAt,
            )
          : null,
    );
  }
}
